File: backend-go/internal/modules/system/service/permission_service.go
package service

import (
	"strings"

	"drip-admin/backend-go/internal/common"
)

func (s *Server) roleCodes(userID common.Int64String) ([]string, error) {
	var roles []SysRole
	err := s.db.Table("sys_role").
		Select("sys_role.*").
		Joins("join sys_user_role on sys_user_role.role_id = sys_role.id").
		Where("sys_user_role.user_id = ? and sys_role.deleted = 0 and sys_role.status = 1", userID.Int64()).
		Find(&roles).Error
	if err != nil {
		return nil, err
	}
	out := make([]string, 0, len(roles))
	for _, role := range roles {
		out = append(out, role.RoleCode)
	}
	return out, nil
}

func (s *Server) permissionCodes(userID common.Int64String) ([]string, error) {
	roles, err := s.roleCodes(userID)
	if err != nil {
		return nil, err
	}
	if containsString(roles, "SUPER_ADMIN") {
		var menus []SysMenu
		if err := s.db.Where("status = 1 and permission_code is not null").Find(&menus).Error; err != nil {
			return nil, err
		}
		return permissionCodesFromMenus(menus), nil
	}
	var menus []SysMenu
	err = s.db.Table("sys_menu").
		Select("distinct sys_menu.*").
		Joins("join sys_role_menu on sys_role_menu.menu_id = sys_menu.id").
		Joins("join sys_user_role on sys_user_role.role_id = sys_role_menu.role_id").
		Where("sys_user_role.user_id = ? and sys_menu.status = 1 and sys_menu.permission_code is not null", userID.Int64()).
		Find(&menus).Error
	if err != nil {
		return nil, err
	}
	return permissionCodesFromMenus(menus), nil
}

func permissionCodesFromMenus(menus []SysMenu) []string {
	seen := map[string]bool{}
	out := make([]string, 0, len(menus))
	for _, menu := range menus {
		if menu.PermissionCode == nil || strings.TrimSpace(*menu.PermissionCode) == "" {
			continue
		}
		if !seen[*menu.PermissionCode] {
			out = append(out, *menu.PermissionCode)
			seen[*menu.PermissionCode] = true
		}
	}
	return out
}

func containsString(values []string, needle string) bool {
	for _, value := range values {
		if value == needle {
			return true
		}
	}
	return false
}

File: backend-python/app/modules/system/service/permission_service.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.modules.system.entity import SysMenu, SysRole, SysRoleMenu, SysUserRole


class PermissionService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def role_codes(self, user_id: int) -> list[str]:
        query = (
            select(SysRole.role_code)
            .join(SysUserRole, SysUserRole.role_id == SysRole.id)
            .where(SysUserRole.user_id == user_id, SysRole.status == 1, SysRole.deleted == 0)
        )
        return list((await self.db.scalars(query)).all())

    async def permission_codes(self, user_id: int) -> list[str]:
        roles = await self.role_codes(user_id)
        if "SUPER_ADMIN" in roles:
            query = select(SysMenu.permission_code).where(
                SysMenu.status == 1, SysMenu.deleted == 0, SysMenu.permission_code.is_not(None)
            )
        else:
            query = (
                select(SysMenu.permission_code)
                .join(SysRoleMenu, SysRoleMenu.menu_id == SysMenu.id)
                .join(SysRole, SysRole.id == SysRoleMenu.role_id)
                .join(SysUserRole, SysUserRole.role_id == SysRole.id)
                .where(
                    SysUserRole.user_id == user_id,
                    SysRole.status == 1,
                    SysRole.deleted == 0,
                    SysMenu.status == 1,
                    SysMenu.deleted == 0,
                    SysMenu.permission_code.is_not(None),
                )
            )
        return sorted(set((await self.db.scalars(query)).all()))

